### PR TITLE
(BOLT-191) Support uploading directories

### PR DESCRIPTION
**Pull request by MikaelSmith**
Implements uploading directories - i.e. files in a directory recursively - with SSH and WinRM.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Doesn't support PCP, I imagine that will need an update to bolt_shim to deserialize a tarball containing multiple files.

**Comment by MikaelSmith** [comment]
There are a couple cases where I just want to upload files and know where they went, but not care about the specific directory. Uploading a temp directory and returning the path could be useful. Bolt uses this internally a number of places.

File: lib/bolt/cli.rb [hunk 1]
```diff
@@ -389,7 +389,7 @@ def pal
       @pal ||= Bolt::PAL.new(config.modulepath, config.hiera_config, config.compile_concurrency)
     end
 
-    def validate_file(type, path)
+    def validate_file(type, path, allow_dir = false)
```
Comment (adreyer) [inline lib/bolt/cli.rb]: If we're using allow_dir do we need to recurse through the directory to make sure it's descendants are readable?

Comment (MikaelSmith) [reply lib/bolt/cli.rb]: That makes sense to me.

**Comment by adreyer** [comment]
We need to add cp -r support for the local transport and minitar support for orchestrator

**Comment by MikaelSmith** [comment]
For orchestrator, it makes sense to me to update `bolt_shim::upload` to accept a new format that notes it's a tarball and where it should be expanded.

**Comment by MikaelSmith** [comment]
Orchestrator supports depends on https://github.com/puppetlabs/puppetlabs-bolt_shim/pull/5. Without it, the fallback behavior will be to write the path out as a tar-gzipped file.

**Comment by MikaelSmith** [comment]
This seems functional, but I think it needs some more testing, particularly with Orchestrator.

File: lib/bolt/transport/orch.rb [hunk 2]
```diff
@@ -114,14 +118,50 @@ def batch_script(targets, script, arguments, options = {}, &callback)
...
+        output&.close
+      end
+
       def batch_upload(targets, source, destination, options = {}, &callback)
-        content = File.open(source, &:read)
+        stat = File.stat(source)
+        content = if stat.directory?
+                    pack(source)
+                  else
+                    File.open(source, &:read)
```
Comment (adreyer) [inline lib/bolt/transport/orch.rb]: is this different than File.read(source)

Comment (MikaelSmith) [reply lib/bolt/transport/orch.rb]: Really doesn't seem like it.

**Pull request merged**